merk/src/merk/chunks.rs

```diff
@@ -26,479 +26,1045 @@
-    /// constructor, the first chunk (the "trunk") will be created.
-    pub fn new(merk: &Merk<S>) -> Result<Self, Error> {
-        let (trunk, has_more) = merk
-            .walk(|maybe_walker| match maybe_walker {
-                Some(mut walker) => walker.create_trunk_proof(),
-                None => Ok((vec![], false)).wrap_with_cost(Default::default()),
-            })
-            .unwrap()?;
-
-        let chunk_boundaries = if has_more {
-            trunk
-                .iter()
-                .filter_map(|op| match op {
-                    Op::Push(Node::KVValueHashFeatureType(key, ..)) => Some(key.clone()),
-                    _ => None,
-                })
-                .collect()
-        } else {
-            vec![]
-        };
-
-        let mut raw_iter = merk.storage.raw_iter();
-        raw_iter.seek_to_first().unwrap();
-
-        Ok(ChunkProducer {
-            trunk,
-            chunk_boundaries,
-            raw_iter,
-            index: 0,
+    /// Creates a new `ChunkProducer` for the given `Merk` instance
+    pub(crate) fn new(merk: &'db Merk<S>) -> Result<Self, Error> {
+        let tree_height = merk
+            .height()
+            .ok_or(Error::ChunkingError(ChunkError::EmptyTree(
+                "cannot create chunk producer for empty Merk",
+            )))?;
+        Ok(Self {
+            height: tree_height as usize,
+            index: 1,
+            merk,
         })
     }
 
     /// Gets the chunk with the given index. Errors if the index is out of
     /// bounds or the tree is empty - the number of chunks can be checked by
     /// calling `producer.len()`.
-    pub fn chunk(&mut self, index: usize) -> Result<Vec<Op>, Error> {
-        if index >= self.len() {
-            return Err(Error::ChunkingError("Chunk index out-of-bounds"));
+    pub fn chunk_with_index(
+        &mut self,
+        chunk_index: usize,
+    ) -> Result<(Vec<Op>, Option<usize>), Error> {
+        let traversal_instructions = generate_traversal_instruction(self.height, chunk_index)?;
+        self.chunk_internal(chunk_index, traversal_instructions)
+    }
+
+    /// Returns the chunk at a given chunk id.
+    pub fn chunk(&mut self, chunk_id: &str) -> Result<(Vec<Op>, Option<String>), Error> {
+        let traversal_instructions = string_as_traversal_instruction(chunk_id)?;
+        let chunk_index = chunk_id_from_traversal_instruction_with_recovery(
+            traversal_instructions.as_slice(),
+            self.height,
+        )?;
+        let (chunk, next_index) = self.chunk_internal(chunk_index, traversal_instructions)?;
+        let index_string = next_index
+            .map(|index| generate_traversal_instruction_as_string(self.height, index))
+            .transpose()?;
+        Ok((chunk, index_string))
+    }
+
+    /// Returns the chunk at the given index
+    /// Assumes index and traversal_instructions represents the same information
```

Comment:
This might be a sign of a new type needed, to ensure two field are in sync to have an invariant